CHANGES.md

```diff
@@ -1,6 +1,421 @@
+o missing contributions from interns (or others?)
+o any missing cross-references?
+o new compiler flags (and developer compiler flags)
+o examples directory
+o man page / compiler flags
+o spellcheck
+o `` vs. ''
+o docs/1.28/...
+o linefeeds
+o TODOs for next time:
+  - initial lowercase
+  - 79 columns
+  - describe changes in terms of user-facing behavior (or devel-facing for
+    those sections)
+  - sort entries by topic / importance
+
+version 1.28.0
+==============
+
+Thirty-first public release of Chapel, September 15, 2022
+
+Highlights (see subsequent sections for further details)
+--------------------------------------------------------
+* significantly improved behavior for numerical ops on small/mixed types
+* added a new 'Communication' module for performing low-level puts/gets
+* expanded the idioms that can run on GPUs and added a new 'GPU' utility module
+* made a number of improvements to 'chpldoc' and 'mason'
+* significantly simplified and improved selection between overloaded routines
+* made a large number of stabilizing language and library improvements
+* modestly reduced average compilation times by using the LLVM support library
+* improved portability to ARM-based Macs (e.g., M1 Macs)
+
+Packaging / Configuration Changes
+---------------------------------
+* `cmake` 3.13.4 or later is now required to build Chapel  
+  (see https://chapel-lang.org/docs/1.28/usingchapel/prereqs.html#readme-prereqs)
+* `chpl` now requires / makes use of the LLVM support library for compilation
+* added `CHPL_LLVM_SUPPORT` to specify the source of the LLVM support library  
+  (see https://chapel-lang.org/docs/1.28/usingchapel/chplenv.html#chpl-llvm-support)
+* made quickstart scripts use `CHPL_LLVM=system` when a good system LLVM exists  
+  (see https://chapel-lang.org/docs/1.28/usingchapel/QUICKSTART.html)
+* made `CHPL_MEM` default to `cstdlib` on Arm-based Macs
+* 'chpldoc' now requires Python 3.7 or later
+
+Syntactic / Naming Changes
+--------------------------
+
+Semantic Changes / Changes to the Chapel Language
+-------------------------------------------------
+* updated `.low`/`.high` on strided ranges/domains to return aligned bounds  
+  (see https://chapel-lang.org/docs/1.28/language/spec/ranges.html#ChapelRange.range.low  
+   and https://chapel-lang.org/docs/1.28/language/spec/ranges.html#ChapelRange.range.high)
+* added support for `[u]int(64)` values to implicitly convert to `real(32)`  
+  (see https://chapel-lang.org/docs/1.28/language/spec/conversions.html#implicit-numeric-and-bool-conversions)
+* added support for `int(w)` values to implicitly convert to `uint(w)`  
+  (see https://chapel-lang.org/docs/1.28/language/spec/conversions.html#implicit-numeric-and-bool-conversions)
+* improved the behavior and result types for mixed-type/-size numerical ops  
+  (see TODO)
+* made `param` numerical ops better match their `var`/`const` equivalents  
+  (e.g., `1:int(8) + 2` now produces `int(64)` rather than `int(8)`)
```

Comment:
This can have

```
  (see https://chapel-lang.org/docs/1.28/language/evolution.html#version-1-28-september-2022)
```

(provided by PR #20653 which isn't merged yet).

Reply:
Great, thanks!  Just to make sure, are you talking about the TODO here (which is what I was imagining putting here once the link was available) or the "param numerical ops" part or both?

Reply:
> * improved the behavior and result types for mixed-type/-size numerical ops  

the PR I merged didn't really talk about this directly. We could add a section about "mixed int/uint ops". In my comment, I was just talking about the "made `param` numerical ops better match their `var`/`const` equivalents " parts.

Reply:
@bradcray @benharsh - can you help with turning this into a PR and merging it?

Here is some draft text for evolution.rst to go right before "Param Expression Behavior":

```
Behavior Differences for Mixes of Signed and Unsigned
*****************************************************

Prior to 1.28, numeric operations applied to a mix of signed and unsigned
could have suprising results by moving the computation from a particular
bit width to another—or by moving it from an integral computation to a
floating point one.

For example:

.. code-block:: chapel

    var myInt:int = 1;
    var myUint:uint = 2;
    var myIntPlusUint = myInt + myUint; // what is the type of `myIntPlusUint`?

Before 1.28, this program would result in compilation error, due to an
error overload of ``operator +`` in the standard library.

Version 1.28 adds the ability for an ``int`` to implicitly convert to
``uint`` and removes the error overload. As a result, the ``uint``
version of ``operator +`` is chosen and that leads to ``myIntPlusUint``
having type ``uint``.

However, a user-defined ``proc plus`` defined for all the numeric types
would resolve to the ``real`` version because ``int`` could not
implicitly convert to ``uint``, but both ``int`` and ``uint`` could
implicitly convert to ``real(64)``. Here is an example that demonstrates
that fact:

.. code-block:: chapel

    proc plus(a: int, b: int)   { return a + b; }
    proc plus(a: uint, b: uint) { return a + b; }
    proc plus(a: real, b: real) { return a + b; }

    var myInt:int = 1;
    var myUint:uint = 2;
    var myIntPlusUint = plus(myInt, myUint);

In 1.27, this would result in ``myIntPlusUint`` having type ``real`` but
in 1.28 it has type ``uint``. This change from integral types to floating
point types could be very surprising.

The next example is similar but uses smaller ``int`` and ``uint`` widths.

.. code-block:: chapel

    var myInt32:int(32) = 1;
    var myUint32:uint(32) = 2;
    var myInt32PlusUint32 = myInt32 + myUint32;

In 1.27, the ``int(64)`` ``+`` operator is chosen (because both
``int(32)`` and ``uint(32)`` can implicitly convert to ``int(64)``).
That results in ``myInt32PlusUint32`` having type ``int(64)`` which would
be surprising to someone who thought they were working with 32 bit
numbers.

In contrast, in 1.28, due to the ability for ``int(32)`` to implicitly
convert to ``uint(32)``, the ``uint(32)`` version of the ``+`` operator
is chosen and ``myInt32PlusUint32`` has type ``uint(32)``.

```

Then

> * improved the behavior and result types for mixed-type/-size numerical ops  

can link to this section.

I didn't get to it in the above draft but it would be nice to also mention that the new behavior allows e.g. `uint` + `uint(8)` to work (see issue #17756). These cases were running afoul of the error overload for int/uint but they work now that the error overload is removed.